### ISSUE TITLE
Hostname check fails for `*.example.com` when `*` contains underscore

### ISSUE THREAD

**Issue by joshua-maros** [openssl/openssl#12566]
Running on Ubuntu 19.10, OpenSSL provided by system, version is 1.1.1c  28 May 2019.

Command to reproduce issue:
```sh
openssl s_client -connect code_cube.gitlab.io:443 -verify_hostname code_cube.gitlab.io
```

The URL https://code_cube.gitlab.io/audio-bench works in a standard web browser. The certificate is valid for `gitlab.io` and `*.gitlab.io`. Removing the underscore in the verification check resolves the issue:

```sh
openssl s_client -connect code_cube.gitlab.io:443 -verify_hostname codecube.gitlab.io
```

**Comment by kroeckx** [comment]
An underscore is not allowed in a hostname (but is allowed in domain name).

**Comment by joshua-maros** [comment]
So in this case, only `github.io` should be provided for verification? If so, how would I be able to automatically determine how to do this? I originally stumbled across this issue by trying to connect to the listed domain using an HTTP request library, the original issue can be seen here: https://github.com/hyperium/hyper/issues/2257. If it is a problem with how the library is being used, I can re-open the original issue with a suggestion on how to change the library.

**Comment by kaduk** [comment]
Well, it seems you're in a tough spot based on the actions of others and not really any fault of your own.
RFC 1034 is quite clear that

> The labels must follow the rules for ARPANET host names.  They must
start with a letter, end with a letter or digit, and have as interior
characters only letters, digits, and hyphen.  There are also some
restrictions on the length.  Labels must be 63 characters or less.

underscore is not one of the listed items "letters, digits, and hyphen" therefor it cannot be present in a compliant host name.  So, if github knew that they would offer the username.github.io service, they should have disallowed underscores in usernames (or project names?) from the start.  They did not, so now you're in this situation where it's in violation of the spec.  I don't think that "browsers do it" is intrinsically a compelling reason for openssl to deviate from the spec, though perhaps there are other reasons that are compelling.

In short, you are asking openssl to violate the spec, so rough edges are unsurprising, and it's not really clear that calling the behavior a "bug" is justified, as frustrating an experience as it may be.

**Comment by joshua-maros** [comment]
It is my understanding that `code_cube.gitlab.io` is still a valid domain name even if it is not a valid hostname, according to @kroeckx 's response as well as [this StackOverflow question](https://stackoverflow.com/questions/2180465/can-domain-name-subdomains-have-an-underscore-in-it). Additionally, [this document on DigiCert's website](https://docs.digicert.com/manage-certificates/retiring-underscores-domain-names/) regarding restrictions on underscores in SSL certificates states the following:

> A wildcard certificate for *.example.com secures example_domain.example.com and _example.domain.example.com.

As stated previously, the issue came up while using an HTTP request library. It is possible I am using the command arguments in a way they were not intended, but I do not have enough experience with OpenSSL to know for sure.

**Comment by kroeckx** [comment]
Digicert's post is just a workaround in case you do have such a
hostname, and the software you're using does not enforce the
rules. They should not have had such a suggestion there, it's just
misleading.

Domain names can contain an underscore, like _443._tcp.example.org,
but they are not hostnames, you can not connect to them.

**Comment by joshua-maros** [comment]
Alright, I think I understand now. Thank you for the explanation.
